presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveMapCastRule.java

```diff
@@ -0,0 +1,124 @@
+{
+    public RemoveMapCastRule(FunctionAndTypeManager functionAndTypeManager)
+    {
+        super(new RemoveMapCastRule.Rewriter(functionAndTypeManager));
+    }
+
+    @Override
+    public boolean isRewriterEnabled(Session session)
+    {
+        return isRemoveMapCastEnabled(session);
+    }
+
+    @Override
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(filterRowExpressionRewriteRule(), projectRowExpressionRewriteRule());
+    }
+
+    private static class Rewriter
+            implements PlanRowExpressionRewriter
+    {
+        private final RemoveMapCastRewriter removeMapCastRewriter;
+
+        public Rewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.removeMapCastRewriter = new RemoveMapCastRewriter(functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewrite(RowExpression expression, Rule.Context context)
+        {
+            return RowExpressionTreeRewriter.rewriteWith(removeMapCastRewriter, expression);
+        }
+    }
+
+    private static class RemoveMapCastRewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+
+        private RemoveMapCastRewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if ((functionResolution.isSubscriptFunction(node.getFunctionHandle()) || functionResolution.isElementAtFunction(node.getFunctionHandle())) && node.getArguments().get(0) instanceof CallExpression
+                    && functionResolution.isCastFunction(((CallExpression) node.getArguments().get(0)).getFunctionHandle())
+                    && ((CallExpression) node.getArguments().get(0)).getArguments().get(0).getType() instanceof MapType) {
+                CallExpression castExpression = (CallExpression) node.getArguments().get(0);
+                RowExpression castInput = castExpression.getArguments().get(0);
+                Type fromKeyType = ((MapType) castInput.getType()).getKeyType();
+                Type fromValueType = ((MapType) castInput.getType()).getValueType();
+                Type toKeyType = ((MapType) castExpression.getType()).getKeyType();
+                Type toValueType = ((MapType) castExpression.getType()).getValueType();
+
+                if (toValueType.equals(fromValueType) && fromKeyType.equals(INTEGER) && toKeyType.equals(BIGINT) && node.getArguments().get(1).getType().equals(BIGINT)) {
```

Comment:
why do we special case the int/bigint pair? shouldn't this work for any cast that is not lossy?

Reply:
Non-exact types as keys is dubious (and also not common) so we don't want to do that.

Reply:
sounds good, if we extract this in a function as suggested above we can also extend this in the future to handle other pairs if needed